Make TryInsert functions within the packages module try to insert first

zeripath opened the pull request:
The TryInsert* functions within the packages models make incorrect assumptions
about transactional isolation within most databases. It is perfectly possible for
a SELECT to return nothing but an INSERT fail with a duplicate in most DBs as it
is only INSERT that the locking occurs.

This PR changes the code to simply try to insert first and if there is an error
then attempt to SELECT from the table. If the SELECT works then the INSERT error
is assumed to have been related to the unique constraint failure.

This technique avoids us having to parse the error returned from the DBs as
these are varied and different.

If the SELECT fails then the INSERT error is returned to the user.

Fix #19586

Signed-off-by: Andrew Thornton <art27@cantab.net>

zeripath commented:
ugh looks like my worries about transactions were right - we're gonna need to use an upsert style...

zeripath commented on models/packages/package_file.go:
```diff
@@ -53,14 +53,10 @@ func TryInsertFile(ctx context.Context, pf *PackageFile) (*PackageFile, error) {
-	if _, err = e.Insert(pf); err != nil {
+	if _, err := e.Insert(pf); err != nil {
+		if has, _ := e.Get(key); has {
+			return pf, ErrDuplicatePackageFile
```
```suggestion
			return key, ErrDuplicatePackageFile
```

zeripath commented:
I'm going to have to mark this WIP as we actually have to use the UPSERT forms otherwise the transactions will fail.

KN4CK3R commented on models/packages/package.go:
```diff
@@ -120,25 +120,30 @@ type Package struct {
+		return nil, err
+	}
+	if n != 0 {
+		return p, nil
```
The insert methods need to return the filled bean. For example it's expected that `ID` has a value in the return value which is not the case here.

zeripath replied:
The code auto fills the ID of the bean.

zeripath commented:
OK I think this is now ready for review.

I do think this kind of upsert style is a far better way of doing this than the current hacky and really quite racy style.

wxiaoguang commented:
Are there dedicated tests for `InsertOnConflictDoNothing` with all databases? Otherwise it would be very difficult to debug or improve it.

wxiaoguang commented on models/packages/package_version.go:
```diff
@@ -37,24 +38,30 @@ type PackageVersion struct {
+		} else if err != nil {
+			return key, err
+		}
+		// This really should never happen and can only happen if this function
+		// is being called outside of a transaction and between the on conflict insert failing
+		// the conlicting item is removed.
+		//
```
There are many `for i := 0; i <= 5; i++` , I do not think it's necessary.

"between the insert failing and conflicting item is removed.": Is there really a case for it? Failing fast without retry is also the correct behavior for all cases.

wxiaoguang replied:
And I believe the retry would case more problems:

Think about 2 threads, they do the same thing:

1. for loop try to insert
2. do something, and insert some related data
2.1 if fail, remove the inserted row
2.2 then, remove other rows related to the inserted row
3. if succeed, over

When thread-B is on step2.1, thread-A is on step1. then thread-A will continue to do something and insert related data

Then thread-A comes to step2, and thread-B comes to 2.2. 

Then thread-B would delete thread-A's data incorrectly.

zeripath replied:
There is no remove.

wxiaoguang replied:
Then why retry? Why would it ever succeed?

wxiaoguang replied:
> I am simply suggesting how it is that this case could possibly happen - otherwise its impossible

I do not think it would ever occur. If it's impossible, the dead code should be removed.

zeripath replied:
It would succeed in the described case.

Imagine the function was called outside a transaction with goroutine A, and a separate goroutine B was also running.

1. Goroutine A tries to INSERT ON CONFLICT but finds that the requested object is in the DB
2. Goroutine B then deletes it.
3. Goroutine A would then try to SELECT * but then finds that the requested object is no longer in the db

What would you prefer to happen then? Would you prefer that we fail? Or would you prefer that we retry?

I wrote retry because I thought reviewers would prefer that we retried. However, if you'd prefer that we simply failed it's very simple to change it.

wxiaoguang replied:
The question is why `Goroutine B then deletes it.` ? 

And even there is a  retrying for 5 times, the problem still happens, Goroutein B could delete the data after Goroutine A's 5-time-retrying. The problem just becomes more complex, no benefit to developers.


It's caller's duty to make sure the data insertion/deletion logic are complete.

wxiaoguang commented on models/db/common.go:
```diff
@@ -20,3 +29,337 @@ func BuildCaseInsensitiveLike(key, value string) builder.Cond {
+		_, _ = sb.WriteString(colNames[0])
+		for _, colName := range colNames[1:] {
+			_, _ = sb.WriteString(",")
+			_, _ = sb.WriteString(colName)
```
I am not sure whether the `colName` is already quoted?

zeripath replied:
Now that I've started quoting everything mssql has broken...

wxiaoguang replied:
I can see a failure in https://drone.gitea.io/go-gitea/gitea/67001/2/17, which is "Error inserting package: error in query: %!s(<nil>), Unsupported type error", it doesn't look like to be related to the column name quoting.

And I really think this new function needs its own test cases, instead of depending on the package tests.

wxiaoguang reviewed (commented):
I can understand what this PR does, but I think it's better to have enough comments .....

If I read the code 3 months ago, I would spend the same time as the first time seeing it.

zeripath commented on models/db/common.go:
```diff
@@ -20,3 +29,337 @@ func BuildCaseInsensitiveLike(key, value string) builder.Cond {
+		case setting.Database.UseSQLite3:
+			_, _ = sb.WriteString(") ON CONFLICT DO NOTHING")
+		case setting.Database.UseMySQL:
+			if autoIncrCol != nil {
+				_, _ = sb.WriteString(") ON DUPLICATE KEY UPDATE ")
+				_, _ = sb.WriteString(autoIncrCol.Name)
+				_, _ = sb.WriteString(" = ")
+				_, _ = sb.WriteString(autoIncrCol.Name)
+			}
```
placating mysql is a magical and wonderful game. I assure that they're both needed.

wxiaoguang replied:
But what's the reason behind it? What makes them both needed ........

zeripath replied:
You realise that they're not used at the same time?

wxiaoguang replied:
Nope, that's too complex, but that's still an unclear case for me about why they both are needed.

`autoIncrCol` is used at many places, I still have the question why `INSERT IGNORE` couldn't be used to replace the `ON DUPLICATE`

wxiaoguang commented on models/db/common.go:
```diff
@@ -20,3 +28,332 @@ func BuildCaseInsensitiveLike(key, value string) builder.Cond {
+		}
+	case setting.Database.UseSQLite3:
+		write(") ON CONFLICT DO NOTHING")
+	case setting.Database.UseMySQL:
+		if autoIncrCol != nil {
+			write(") ON DUPLICATE KEY UPDATE ", quote(autoIncrCol.Name), " = ", quote(autoIncrCol.Name))
+		}
```
Sorry to bother, but why `INSERT IGNORE` doesn't work for `autoIncrCol != nil`?

If you meant to do it for `LastInsertId`, I think it doesn't work.

Otherwise I really can not understand its purpose.

zeripath replied:
Why don't you try it. When I wrote this code it required the ON DUPLICATE KEY UPDATE and not the IGNORE

wxiaoguang replied:
BECAUSE I HAVE TRIED


![image](https://user-images.githubusercontent.com/2114189/217302383-264e0f4c-123c-4708-9f0c-db151bdac98b.png)

wxiaoguang replied:
Table:

```sql
CREATE TABLE `t` (
  `id` int NOT NULL AUTO_INCREMENT,
  `k` varchar(100),
  `v` varchar(100) DEFAULT '',
  `i` int DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `k` (`k`)
) ENGINE=InnoDB AUTO_INCREMENT=9
```

wxiaoguang replied:
What you really need is this:

`update id=last_insert_id(id)`


```
mysql> insert into t (k) values ('1') on duplicate key update id=last_insert_id(id);
Query OK, 0 rows affected (0.00 sec)

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                1 |
+------------------+
1 row in set (0.00 sec)

```

zeripath replied:
a) I would have thought that last_insert_id() requires that there are no other id insertions in the meantime.
b) id = id works
c) IGNORE is only used when there is no autoincrement ID present in the table where there is no id to update.
d) I've written some tests now.

wxiaoguang replied:
```
mysql> insert into t (k) values ('a');
Query OK, 1 row affected (0.00 sec)

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                1 |
+------------------+
1 row in set (0.00 sec)

mysql> insert into t (k) values ('b');
Query OK, 1 row affected (0.00 sec)

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                2 |
+------------------+
1 row in set (0.00 sec)

mysql> insert into t (k) values ('a') on duplicate key update id=id;
Query OK, 0 rows affected (0.00 sec)

!!!!!!! HERE You are inserting duplicated `a` and want to see `id=1`, BUT:

mysql> select last_insert_id();
+------------------+
| last_insert_id() |
+------------------+
|                2 |
+------------------+
1 row in set (0.00 sec)

mysql> select * from t where id=2;
+----+------+------+------+
| id | k    | v    | i    |
+----+------+------+------+
|  2 | b    |      | NULL |
+----+------+------+------+
1 row in set (0.00 sec)

```

wxiaoguang commented on models/db/common.go:
```diff
@@ -20,3 +28,360 @@ func BuildCaseInsensitiveLike(key, value string) builder.Cond {
+		return true, nil
+	}
+
+	res, err := e.Exec(values...)
```
Does it need to be `Exec(insertArgs...)` here? Just a guess because there is a 
`Query(insertArgs...)` above

zeripath replied:
yup mistake when I changed things

wxiaoguang commented:
If I understand correctly, the test cases haven't cover the behavior of comment:

> // This function will update the ID of the provided bean if there is an insertion

wxiaoguang commented:
If I understand correctly, there are still some problems in code:

1. the `has` is misleading. 
    * If insertion succeeds, it returns `true, nil`, if insertion fails, it returns `false, nil`
    * So the meaning of `bool` in `(bool, error)` is: whether the insertion succeeds, not about  whether the record existed before.

2. `assert.NotEqual(t, int(0), int64(0))` always succeeds, But the `ID` is a `int64`, so you must use `NotEqualValues` or `NotEmpty`, otherwise the test code like `assert.NotEqual(t, 0, toInsert.ID)` always succeeds.

3. There is only one `ON CONFLICT DO NOTHING` for SQLite3, I do not understand why it would return the ID for the duplicated insertion.

4. The MySQL is highly like to use `update id=last_insert_id(id)`, but not `update id=id`, the second one doesn't seem the proper way to update the `LAST_INSERT_ID` for MySQL.

lunny commented:
Looks like things become too complex to difficult to review.

zeripath commented:
> If I understand correctly, the test cases haven't cover the behavior of comment:

It feels a bit like you're being deliberately obstructive and I don't quite understand why you're being this way.

This might be a language issue - but there are better ways to say I think you should add a line that checks if the ID remains at 0.

To which I might reply why would we care and why are you being so bloody difficult about it? That the ID remains unchanged is irrelevant. Fine I'll add it but honestly it's just making the test less clear and less easy to read for zero real world benefit. 

It feels like you're targeting me for some reason and I don't understand why. You left maintainers - was it because of me?

> If I understand correctly, there are still some problems in code:
> 
> 1. the `has` is misleading.
>    
>    * If insertion succeeds, it returns `true, nil`, if insertion fails, it returns `false, nil`
>    * So the meaning of `bool` in `(bool, error)` is: whether the insertion succeeds, not about  whether the record existed before.

I was using `has` as in «has been inserted» but I agree `inserted` would be better. Frankly its minor and you should just say use `inserted` instead of `has` instead of making some major issue about it.

> 2. `assert.NotEqual(t, int(0), int64(0))` always succeeds, But the `ID` is a `int64`, so you must use `NotEqualValues` or `NotEmpty`, otherwise the test code like `assert.NotEqual(t, 0, toInsert.ID)` always succeeds.

Again just make a suggestion. It's a good catch that I forgot about it but your tone is frankly insulting.

> 3. There is only one `ON CONFLICT DO NOTHING` for SQLite3, I do not understand why it would return the ID for the duplicated insertion.

I don't understand your question?

There are two cases:
1. INSERT proceeds because there is no conflict. The SQLite driver will return the inserted id because SQLite always does not.
2. INSERT fails because there is a conflict.

In the second case we do not get an ID. Nor do we attempt to use it because 0 rows are affected. 

> 4. The MySQL is highly likely needed to use `update id=last_insert_id(id)`, but not `update id=id`. The second one doesn't seem the proper way to update the `LAST_INSERT_ID` for MySQL.

This code is not attempting to update the `id` in fact it is deliberately not attempting to cause an update to the `id` - without the `ON DUPLICATE KEY UPDATE id = id` there is no `ON CONFLICT DO NOTHING`.

wxiaoguang commented:
> It feels a bit like you're being deliberately obstructive and I don't quite understand why you're being this way.

I am not obstructive, but really want to help, I have tested a lot on my side.

I have tested SQLite3 and MySQL. I do not think your code could work for these cases, and the tests were incorrect.

For SQLite, you need: `insert ... on conflict update set id=id returning id`

For MySQL, you need: `insert ... on duplicate key update id=last_insert_id(id)`

That's all.

>  but your tone is frankly insulting.

Sorry I didn't mean that. But just want make the problem clear. Could you suggest about how to express the problem more gently.

We can have a private chat with the SQL problems later.

zeripath commented:
We don't use or get the ID when the INSERT fails because there is a conflict because `OnConflictDoNothing`.

> I have tested SQLite3 and MySQL. I do not think your code could work for some cases, and the tests were incorrect.
> 
> For SQLite, `insert ... on conflict update set id=id returning id` (`do nothing` doesn't work and won't return the id)

Clearly I need to separate out the code paths here.

There is no `RETURNING` in the SQLite code. The `RETURNING` is only used in Postgres.

SQLite will always return the last inserted ID as part of its driver so we don't need to do that.

If there is a conflict we don't use it - we do nothing.

> For MySQL, `insert ... on duplicate key update id=last_insert_id(id)`

No. As far as I can see that would cause every conflict to cause an update of the primary key (or fail.)

The code as written: `ON DUPLICATE KEY UPDATE id = id` makes the UPDATE a no-op so that no row is changed i.e. NOTHING happens when there is a conflict.

Again the code is not attempting to return the ID when there is a conflict.

> And the test has to cover: insert(keyA), insert(keyB), insert(keyA) and assert getting A-id

See above, the code does not promise to get the ID when there is a conflict. It promises to do nothing.

> There is still no comment about why `INSERT IGNORE` doesn't work for MySQL in such case. (there was just a joking for the first question [#21063 (comment)](https://github.com/go-gitea/gitea/pull/21063#discussion_r1098796187) , I am not good enough to guess).

Using INSERT IGNORE didn't work in my testing when there was a primary key with autoincrement - and if I didn't instead use the ON DUPLICATE KEY UPDATE no-op it would error out. This may have been warnings being returned rather than errors.

It seems however now that `INSERT IGNORE` will just work by itself but I do wonder if the warnings that IGNORE could cause failures or complaints by users.
 
> And, maybe I really misunderstood something -- about the design or about the logic (no comment nor test at the beginning), I do not know. Maybe the divergence starts here [#21063 (comment)](https://github.com/go-gitea/gitea/pull/21063#discussion_r1098526327): "The code auto fills the ID of the bean in the same way that xorm's Insert() does.": The bean should has the ID filled when there is no error.

Ah. Perhaps that is the problem. 

I'm not promising to get the ID when there is no insert. 

The function has comment on it that says it will update the ID of the provided bean if it is inserted. It doesn't say it will do that when it's not inserted - and in fact it will NOT do that - instead on conflict it will do nothing as the function name suggests.

```
// InsertOnConflictDoNothing will attempt to insert the provided bean but if there is a conflict it will not error out
// This function will update the ID of the provided bean if there is an insertion
// This does not do all of the conversions that xorm would do automatically but it does quite a number of them
// once xorm has a working InsertOnConflictDoNothing this function could be removed.
```

There were always tests - just not specific ones. The package repositories code has good coverage of this function. As requested though I've added some specific tests and covered and fixed some more edge cases which should make the function usable in other contexts too.

wxiaoguang commented:
> Using INSERT IGNORE didn't work in my testing when there was a primary key with autoincrement - and if I didn't instead use the ON DUPLICATE KEY UPDATE no-op it would error out. This may have been warnings being returned rather than errors.
> 
> It seems however now that `INSERT IGNORE` will just work by itself but I do wonder if the warnings that IGNORE could cause failures or complaints by users.

From my knowledge, I always use `INSERT IGNORE` for all cases  and  haven't got any problem. That's why the question first came ....  because in my mind:
* `INSERT IGNORE` should  be fine  for all cases, I  never  thought any user  would be surprised or affected by the warning. If  XORM is affected  by the warning, that's  another story. (if there was a comment for it, I  could also get the point as early as  possible).
* Then after being told that the  `IGNORE`  doesn't work for autoincrement, I thought that  since the `insert  on dup update id=id` trick is used, it must be done by purpose to get  the conflicted ID according to  another comment, then the discussion happens.

Thank you for letting me  know more  details.

Anyway, it's clear for me now. I think this PR is good  to have,  maybe it could help  to remove the workaround mutex for the package code after it gets merged:
* #21862
* https://github.com/go-gitea/gitea/pull/21862/files#diff-239d7f1fa93717ce75c12a91af1cc9f9d585993f7a1da9b5a507606689df4994R39-R41

zeripath commented:
Ah yes that mutex can be removed too.

Apologies for the confusion earlier - it just felt we were going round and round in circles for no good reason and I couldn't understand why.

lunny commented:
Thanks for your hard-working @zeripath . It really should be a part of XORM. :)

wxiaoguang commented:
How about taking this PR now, and move to XORM in the future?

The "workaround mutex" really seems fragile, if there is a chance, we could use a DB insert-ignore instead of that in-progress mutex for a stable release.

lunny commented on tests/integration/db_common_test.go:
```diff
@@ -0,0 +1,220 @@
+func TestInsertOnConflictDoNothing(t *testing.T) {
+	defer tests.PrepareTestEnv(t)()
+
+	ctx := db.DefaultContext
```
Should we have a parallel test?

lunny commented:
I think the key problem is the package property table have no unique keys. per https://github.com/go-gitea/gitea/pull/30335#issuecomment-2051982310
